pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/PartitionedConsumerImpl.java

```diff
@@ -350,9 +384,9 @@ private ConsumerConfiguration getInternalConsumerConfig() {
         if (conf.getAckTimeoutMillis() != 0) {
             internalConsumerConfig.setAckTimeout(conf.getAckTimeoutMillis(), TimeUnit.MILLISECONDS);
         }
-        internalConsumerConfig.setMessageListener((consumer, msg) -> {
-            if (msg != null) {
-                messageReceived(msg);
+        internalConsumerConfig.setReceiveListener((consumer) -> {
+            if (receiveRunning.compareAndSet(false, true)) {
+                receiveMessagesFromInternalConsumers(consumers.indexOf((ConsumerImpl) consumer));
```

Comment:
Yes, do you recommend another approach?
We could make  `receiveMessagesFromInternalConsumers` not receive an index and just start from 0 and iterating through consumers.